### PR TITLE
feat(core) : reducing locks coverage to avoid potential deadlocks

### PR DESCRIPTION
**Pull request by Prateeknandle**
**Purpose of PR?**:

Fixes #

**Does this PR introduce a breaking change?**

**If the changes in this PR are manually verified, list down the scenarios covered:**:

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_

**Checklist:**
- [ ] Bug fix. Fixes #<issue number>
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: KubeArmor/core/kubeUpdate.go [hunk 1]
```diff
@@ -516,13 +516,20 @@ func (dm *KubeArmorDaemon) UpdateEndPointWithPod(action string, pod tp.K8sPod) {
...
-			for nidx < len(endpoints) && idx < len(dm.EndPoints) {
-				if pod.Metadata["namespaceName"] == dm.EndPoints[idx].NamespaceName && pod.Metadata["podName"] == dm.EndPoints[idx].EndPointName && kl.ContainsElement(endpoints, dm.EndPoints[idx].ContainerName) {
+			for nidx < len(endpoints) && idx < endPointsLength {
+				dm.EndPointsLock.RLock()
+				endPoint := dm.EndPoints[idx]
+				dm.EndPointsLock.RUnlock()
+				if pod.Metadata["namespaceName"] == endPoint.NamespaceName && pod.Metadata["podName"] == endPoint.EndPointName && kl.ContainsElement(endpoints, endPoint.ContainerName) {
+					dm.EndPointsLock.Lock()
 					dm.EndPoints[idx] = endpoints[nidx]
+					dm.EndPointsLock.Unlock()
```
Comment (daemon1024) [inline KubeArmor/core/kubeUpdate.go]: Let's not do this, it's unnecessary. It's fine to have the lock outside the loop and keep it like earlier.

File: KubeArmor/core/kubeUpdate.go [hunk 2]
```diff
@@ -360,19 +362,17 @@ func (dm *KubeArmorDaemon) UpdateEndPointWithPod(action string, pod tp.K8sPod) {
...
-		dm.EndPointsLock.Lock()
+		dm.EndPointsLock.RLock()
 		for _, endPoint := range dm.EndPoints {
 			if pod.Metadata["namespaceName"] == endPoint.NamespaceName && pod.Metadata["podName"] == endPoint.EndPointName {
 				endpoints = append(endpoints, endPoint)
 			}
 		}
-		dm.EndPointsLock.Unlock()
+		dm.EndPointsLock.RUnlock()
 		if len(endpoints) == 0 {
```
Comment (daemon1024) [inline KubeArmor/core/kubeUpdate.go:377]: Is the only goal for the above for loop is just to check if it's not 0? it should be fine to break out of loop once we add one endpoint

**Review by daemon1024 — COMMENTED** [review]
Initial overview

**Review by daemon1024 — DISMISSED** [review]

**Review by rootxrishabh — DISMISSED** [review]
Basic sanity successful - 

```
rootxrishabh@fedora:~/KubeArmor/KubeArmor/build$ karmor logs 
local port to be used for port forwarding kubearmor-relay-66999cb886-scthr: 32785 
Created a gRPC client (localhost:32785)
Checked the liveness of the gRPC server
Started to watch alerts
== Alert / 2024-11-13 14:50:33.957179 ==
ClusterName: default
HostName: gke-rishabh-cluster-default-pool-7ceba30f-fpcc
NamespaceName: default
PodName: nginx-bf5d5cf98-2s97v
Labels: app=nginx
ContainerName: nginx
ContainerID: d74ce140d48a586c5d6c55b5586356a635d6d3975e15ebe7a7aa86e0d238544d
ContainerImage: docker.io/library/nginx:latest@sha256:bc5eac5eafc581aeda3008b4b1f07ebba230de2f27d47767129a6a905c84f470
Type: MatchedPolicy
PolicyName: block-pkg-mgmt-tools-exec
Severity: 1
Source: /usr/bin/bash
Resource: /usr/bin/apt
Operation: Process
Action: Block
Data: lsm=SECURITY_BPRM_CHECK
Enforcer: BPFLSM
Result: Permission denied
Cwd: /
HostPID: 99824
HostPPID: 99799
Owner: map[Name:nginx Namespace:default Ref:Deployment]
PID: 43
PPID: 99799
ParentProcessName: /usr/bin/bash
ProcessName: /usr/bin/apt
UID: 0


rootxrishabh@fedora:~/KubeArmor/KubeArmor/build$ k exec -it nginx-bf5d5cf98-2s97v -n default  -- bash
root@nginx-bf5d5cf98-2s97v:/# apt
bash: /usr/bin/apt: Permission denied

Containers:
  kubearmor:
    Container ID:  containerd://7c30be7cf790666a4bb38648d71c0544712066dbf6d6164b382a0dafe336ab47
    Image:         rootxrishabh/kubearmor:deadlock
    Image ID:      docker.io/rootxrishabh/kubearmor@sha256:310790aa69a419da76e16a4a09d0d2e779a3ed0082df6fb4068058db5c25f5c9

```

Environment - 
 - K8s -> GKE
 - Container Runtime -> containerd://1.7.22
 - Kernel Version -> 6.1.100+
 - Operating system -> Container-Optimized OS from Google

**Comment by rksharma95** [comment]
race conditions found, @Prateeknandle can you check again? 
[kubearmor_race.log](https://github.com/user-attachments/files/17742915/kubearmor_race.log)

**Comment by DelusionalOptimist** [comment]
@Prateeknandle this is a must for stable release. Can you please check @rksharma95's comments?
cc @daemon1024

**Review by daemon1024 — APPROVED** [review]

**Review by DelusionalOptimist — APPROVED** [review]
CITY

**Pull request merged**